pkg/offline/offline.go

```diff
@@ -157,6 +157,146 @@ func WithQueue(filepath string, syncLimit int) (heartbeat.HandleOption, error) {
+			log.Debugln("abort execution, as there are no queued heartbeats ready for sending")
+
+			return nil
+		}
+
+		results, err := next(hh)
+		if err != nil {
+			requeueErr := pushHeartbeats(filepath, hh)
+			if requeueErr != nil {
+				log.Errorf("failed to push heatbeats after api error to queue: %s", requeueErr)
+			}
+
+			return err
+		}
+
+		return handleResults(filepath, results, hh)
+	}
+}
+
+func handleResults(filepath string, results []heartbeat.Result, hh []heartbeat.Heartbeat) error {
+	var (
+		err               error
+		withInvalidStatus []heartbeat.Heartbeat
+	)
+
+	// push heartbeats with invalid result status codes to queue
+	for n, result := range results {
+		if n >= len(hh) {
+			log.Warnln("results from api not matching heartbeats sent")
+			break
+		}
+
+		if result.Status != http.StatusCreated &&
+			result.Status != http.StatusAccepted &&
+			result.Status != http.StatusBadRequest {
+			withInvalidStatus = append(withInvalidStatus, hh[n])
+		}
+	}
+
+	if len(withInvalidStatus) > 0 {
+		log.Debugf("pushing %d heartbeat(s) with invalid result to queue", len(withInvalidStatus))
+
+		err = pushHeartbeats(filepath, withInvalidStatus)
+		if err != nil {
+			log.Errorf("failed to push heatbeats with invalid status to queue: %s", err)
+		}
+	}
+
+	// handle leftover heartbeats
+	leftovers := len(hh) - len(results)
+	if leftovers > 0 {
+		log.Warnf("Missing %d results from api.", leftovers)
+
+		start := len(hh) - leftovers
+
+		err = pushHeartbeats(filepath, hh[start:])
+		if err != nil {
+			log.Errorf("failed to push leftover heatbeats to queue: %s", err)
+		}
+	}
+
+	return err
+}
+
+func popHeartbeats(filepath string, syncLimit int) ([]heartbeat.Heartbeat, error) {
+	db, err := bolt.Open(filepath, 0600, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open db connection: %s", err)
+	}
+
+	defer db.Close()
+
+	tx, err := db.Begin(true)
+	if err != nil {
+		return nil, fmt.Errorf("failed to start db transaction: %s", err)
+	}
+
+	queue, err := NewQueue(tx)
+	if err != nil {
+		return nil, fmt.Errorf("failed initialize new queue: %s", err)
+	}
+
+	queued, err := queue.PopMany(syncLimit)
+	if err != nil {
+		errrb := tx.Rollback()
+		if errrb != nil {
+			log.Errorf("failed to rollback transaction: %s", errrb)
+		}
+
+		return nil, fmt.Errorf("failed to pop heartbeat(s) from queue: %s", err)
+	}
+
+	if err = tx.Commit(); err != nil {
+		return nil, fmt.Errorf("failed to commit db transaction: %s", err)
+	}
+
+	return queued, nil
+}
+
+func pushHeartbeats(filepath string, hh []heartbeat.Heartbeat) error {
+	db, err := bolt.Open(filepath, 0600, nil)
+	if err != nil {
+		return fmt.Errorf("failed to open db connection: %s", err)
+	}
+
+	defer db.Close()
+
+	tx, err := db.Begin(true)
+	if err != nil {
+		return fmt.Errorf("failed to start db transaction: %s", err)
+	}
+
+	queue, err := NewQueue(tx)
+	if err != nil {
+		return fmt.Errorf("failed initialize new queue: %s", err)
+	}
+
+	err = queue.PushMany(hh)
```

Comment:
Can we have some retry here? I'm thinking if there are any errors due to another process having the global boltdb lock, we should retry so heartbeats aren't lost.

Reply:
Added a retry. Check new commits :+1: